### ISSUE TITLE
Differences from games created with plain cocos2d-x

### ISSUE THREAD

**Issue by jonnyijapan** [NatWeiss/RapidGame#23]
I'm still not very familiar with cocos2d-x itself, but it seems that the game template created by RapidGame differs from the template created by cocos2d-x, specifically the one I created with v3.5.
The project I work on now, is a template I created recently with RapidGame and it uses cocos2d-x v3.6 so the difference might stem from that version change, however:

The iOS AppController.mm is different in the app state change hooks like the one below. In cocos2d-x v3.5 it actually does nothing "anymore":

```
- (void)applicationWillResignActive:(UIApplication *)application {
    /*
     Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
     Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
     */
     //We don't need to call this method any more. It will interupt user defined game pause&resume logic
    /* cocos2d::Director::getInstance()->pause(); */
}
```

With rapidgame-cocos2d v3.6 it looks like it does that code "again":

```
-(void) applicationWillResignActive:(UIApplication*)application
{
	Director::getInstance()->pause();
}
```

Which is the correct one really?

On another note, I noticed that "the way" to create scenes with RapidGame-cocos2d3.6 is to just subclass cocos2d::Scene and then do replaceScene etc with that. With Cocos2dx3.5 I always subclassed Layer or similar, then had a createScene which made a Scene and then added the layer as a child etc. I think the RapidGame way is more sleak of course and I'm just changing my scenes around at the moment to work with that - because if I did not do that I encountered a serious error when transitioning between scenes - cocos2dx complained about nodes that were still active at the time they were deallocated, or something like that. That problem only went away when I changed the way scenes are subclassed as described above. Any light?

**Comment by NatWeiss** [comment]
Regarding your first question, I do not know which was is currently correct. The cocos2d-x team makes many changes with every release and it is difficult to keep up with what is the correct way to do what. If I had to guess I would say that mimicking the cocos2d-x templates / test apps is the right way to go.

Regarding your second question, one of the changes that cocos2d-x made recently (like in the last year or two) was to make everything a `Node`. So both `Scene` and `Layer` are a subclass of `Node`. So it really shouldn't matter which one you use as long as your are handling the memory correctly you won't get any errors about nodes still being active. `addChild()` calls `retain()` so remember to use `release()` if you use `new` combined with `addChild()`.

**Comment by jonnyijapan** [comment]
I went at back at looked at this again, because using replaceScene(new Myscene) instead of `replaceScene(Myscene::createScene)` like in the template of cocos2d-x seems to indeed not deallocate the previous scene properly, at least not in my game. For some reason it seems to properly only deallocate the initial scene on first replaceScene() call, but on later calls destructors are no longer called. So this would mean that we have to call delete like you say - but on the *scenes*. Which I think a game dev cannot do, this should be handled by cocos2d. Which leeds me to think that we have to stick to the createScene pattern used in recent cocos2d (at least 3.5 through 3.7).
Problem here if I use createScene with the libs compiled with RapidGame I get hard crashes when calling replaceScene with this error:

"""
Assert failed: Node still marked as running on node destruction! Was base class onExit() called in derived class onExit() implementations?
Assertion failed: (!_running), function ~Node, file /Users/jonny/Library/Developer/RapidGame/src/cocos2d-js/frameworks/js-bindings/cocos2d-x/cocos/2d/CCNode.cpp, line 212.
"""


I have stripped the scene of any weird stuff and am still trying to debug this, but have no luck so far. The above crash happens already BEFORE onExitTransitionDidStart is even called, but after replaceScene().
